### PR TITLE
Operator fee timelocks

### PR DESCRIPTION
**Pull request by vadiminc**

### CODE REVIEW

File: contracts/SSVNetwork.sol [hunk 1]
```diff
@@ -53,36 +59,76 @@ contract SSVNetwork is OwnableUpgradeable, ISSVNetwork {
...
+    function __SSVNetwork_init_unchained(
+        IERC20 token_,
+        uint64 operatorMaxFeeIncrease_,
+        uint256 declareOperatorFeePeriod_,
+        uint256 executeOperatorFeePeriod_
+    ) internal onlyInitializing {
         _token = token_;
+        _updateOperatorFeeIncreaseLimit(operatorMaxFeeIncrease_);
+        _updateDeclareOperatorFeePeriod(declareOperatorFeePeriod_);
+        _updateDeclareOperatorFeePeriod(declareOperatorFeePeriod_);
```
Comment (andrew-blox) [inline contracts/SSVNetwork.sol:116]: These lines are identical

File: contracts/SSVNetwork.sol [hunk 2]
```diff
@@ -104,13 +150,56 @@ contract SSVNetwork is OwnableUpgradeable, ISSVNetwork {
         emit OperatorRemoved(operatorId);
     }
 
-    function updateOperatorFee(uint64 operatorId, uint64 fee) external {
+    function declareOperatorFee(uint64 operatorId, uint64 fee) onlyOperatorOwnerOrContractOwner(operatorId) external {
```
Comment (andrew-blox) [inline contracts/SSVNetwork.sol:153]: In general with uint64 highest fee possible is ~18 .  We need to add shrink and expand functions (affects many other places as well)

File: test/operators/fee.ts [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+    await helpers.registerOperators(2, 1, helpers.CONFIG.minimalOperatorFee);
+  });
+
+  it('Declare fee success emits OperatorFeeDeclaration event', async () => {
+    await expect(ssvNetworkContract.connect(helpers.DB.owners[2]).declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee +  helpers.CONFIG.minimalOperatorFee / 10))
+      .to.emit(ssvNetworkContract, 'OperatorFeeDeclaration');
+  });
+
+  it('Declare fee success as contract owner', async () => {
+    await ssvNetworkContract.declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee +  helpers.CONFIG.minimalOperatorFee / 10);
```
Comment (andrew-blox) [inline test/operators/fee.ts:20]: - add gas track with register operator param
- There are many places with this, the basic rule is if it doesnt have emit or revert it should be in gas tracking.

@lior-blox 
- Maybe its best that all write functions have a helper inside contract-helpers.ts file and in this function is built in gas tracking.  
- Also maybe not to write all the dao functions inside contract-helpers.ts, but instead to wrap them with gasTrack inside the test (to not clutter helpers file with functions that will only be tested once).

File: test/operators/fee.ts [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+      .to.be.revertedWith('FeeTooLow');
+  });
+
+  it('Declare fee fails fee exceeds increase limit', async () => {
+    await expect(ssvNetworkContract.connect(helpers.DB.owners[2]).declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee +  helpers.CONFIG.minimalOperatorFee / 5))
+      .to.be.revertedWith('FeeExceedsIncreaseLimit');
+  });
+
+  it('Cancel declared fee success emits DeclaredOperatorFeeCancelation event', async () => {
+    await ssvNetworkContract.declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee + helpers.CONFIG.minimalOperatorFee / 10);
```
Comment (andrew-blox) [inline test/operators/fee.ts:39]: add gas track with register operator param

File: test/operators/fee.ts [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+
+    await expect(ssvNetworkContract.connect(helpers.DB.owners[2]).cancelDeclaredOperatorFee(1))
+      .to.emit(ssvNetworkContract, 'DeclaredOperatorFeeCancelation');
+  });
+
+  it('Cancel declared fee success as contract owner', async () => {
+    await ssvNetworkContract.declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee + helpers.CONFIG.minimalOperatorFee / 10);
+
+    await ssvNetworkContract.cancelDeclaredOperatorFee(1);
```
Comment (andrew-blox) [inline test/operators/fee.ts:48]: For both add gas track with register operator param

File: test/operators/fee.ts [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+  });
+
+  it('Cancel declared fee fails no pending request', async () => {
+    await expect(ssvNetworkContract.cancelDeclaredOperatorFee(1))
+      .to.be.revertedWith('NoPendingFeeChangeRequest');
+  });
+
+  it('Cancel declared fee fails no owner', async () => {
+    await ssvNetworkContract.declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee + helpers.CONFIG.minimalOperatorFee / 10);
```
Comment (andrew-blox) [inline test/operators/fee.ts:57]: add gas track with register operator param

File: test/operators/fee.ts [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+
+    await expect(ssvNetworkContract.connect(helpers.DB.owners[1]).cancelDeclaredOperatorFee(1))
+      .to.be.revertedWith('CallerNotOwner');
+  });
+
+  it('Execute declared fee success emits OperatorFeeExecution event', async () => {
+    await ssvNetworkContract.declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee + helpers.CONFIG.minimalOperatorFee / 10);
```
Comment (andrew-blox) [inline test/operators/fee.ts:64]: add gas track with register operator param

File: test/operators/fee.ts [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+    await progressTime(helpers.CONFIG.declareOperatorFeePeriod);
+    await expect(ssvNetworkContract.connect(helpers.DB.owners[2]).executeOperatorFee(1))
+      .to.emit(ssvNetworkContract, 'OperatorFeeExecution');
+  });
+
+  it('Execute declared fee success as contract owner', async () => {
+    await ssvNetworkContract.declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee + helpers.CONFIG.minimalOperatorFee / 10);
+    await progressTime(helpers.CONFIG.declareOperatorFeePeriod);
+    await ssvNetworkContract.executeOperatorFee(1);
```
Comment (andrew-blox) [inline test/operators/fee.ts:73]: For both contract calls add gas track with register operator param

File: test/operators/fee.ts [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+  });
+
+  it('Execute declared fee fee fails no owner', async () => {
+    await ssvNetworkContract.declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee + helpers.CONFIG.minimalOperatorFee / 10);
```
Comment (andrew-blox) [inline test/operators/fee.ts:77]: add gas track with register operator param

File: test/operators/fee.ts [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+      .to.be.revertedWith('CallerNotOwner');
+  });
+
+  it('Execute declared fee fails no pending request', async () => {
+    await expect(ssvNetworkContract.executeOperatorFee(1))
+      .to.be.revertedWith('NoPendingFeeChangeRequest');
+  });
+
+  it('Execute declared fee fails not within timeframe', async () => {
+    await ssvNetworkContract.declareOperatorFee(1, helpers.CONFIG.minimalOperatorFee + helpers.CONFIG.minimalOperatorFee / 10);
```
Comment (andrew-blox) [inline test/operators/fee.ts:89]: add gas track with register operator param

File: contracts/SSVNetwork.sol [hunk 11]
```diff
@@ -100,13 +151,56 @@ contract SSVNetwork is OwnableUpgradeable, ISSVNetwork {
...
+
+    function executeOperatorFee(uint64 operatorId) onlyOperatorOwnerOrContractOwner(operatorId) external {
+        OperatorFeeChangeRequest memory feeChangeRequest = _operatorFeeChangeRequests[operatorId];
+
+        if(feeChangeRequest.fee == 0) {
+            revert NoPendingFeeChangeRequest();
+        }
+
+        if(block.timestamp < feeChangeRequest.approvalBeginTime || block.timestamp > feeChangeRequest.approvalEndTime) {
+            revert ApprovalNotWithinTimeframe();
```
[the remark this answers is not in the document]
Comment (vadiminc) [reply contracts/SSVNetwork.sol:198]: @adamzigdon ?

File: contracts/SSVNetwork.sol [hunk 12]
```diff
@@ -100,13 +151,56 @@ contract SSVNetwork is OwnableUpgradeable, ISSVNetwork {
...
+
+        if (fee.shrink() > operator.fee * (10000 + _operatorMaxFeeIncrease) / 10000) {
+            revert FeeExceedsIncreaseLimit();
+        }
+
+        /*
+        if (operatorFee <= operator.fee) {
+            _updateOperatorFeeUnsafe(operatorId, operatorFee);
+        } else {
+        */
```
[the remark this answers is not in the document]
Comment (vadiminc) [reply contracts/SSVNetwork.sol:169]: @adamzigdon ? @lior-blox ?

File: contracts/SSVNetwork.sol [hunk 13]
```diff
@@ -555,6 +691,26 @@ contract SSVNetwork is OwnableUpgradeable, ISSVNetwork {
         return balance < LIQUIDATION_MIN_BLOCKS * (_burnRatePerValidator(_extractOperators(operatorIds)) + _networkFee) * validatorCount;
     }
 
+    function _updateOperatorFeeUnsafe(uint64 operatorId, uint64 fee) private {
```
[the remark this answers is not in the document]
Comment (vadiminc) [reply contracts/SSVNetwork.sol:694]: @adamzigdon ?

File: test/operators/fee.ts (hunk 3 above)
Comment (lior-blox) [reply test/operators/fee.ts:20]: agree

**Pull request merged**